server/raft.go

```diff
@@ -4282,9 +4297,35 @@ func (n *raft) switchToLeader() {
 	n.lxfer = false
 	n.updateLeader(n.id)
 	n.switchState(Leader)
+
+	// Wait for messages to be applied if we've stored more, otherwise signal immediately.
+	if n.pindex > n.applied {
+		n.aflr = n.pindex
+	} else {
+		n.signalAppliedFloor()
+	}
 	n.Unlock()
 
 	if sendHB {
 		n.sendHeartbeat()
 	}
 }
+
+// Will signal that all messages that were not yet applied when becoming leader, have been applied since.
+func (n *raft) signalAppliedFloor() {
+	// We don't care about values that have not been consumed (transitory states),
+	// so we dequeue any state that is pending and push the new one.
+	for {
+		select {
```

Comment:
I would just do a simple select with default being a no-op, meaning the chan is full so someone will pick it up.

Reply:
We use this pattern a bit throughout the code base.

Reply:
Done